src/charm.py

```diff
@@ -64,6 +66,92 @@ def evaluate(self, csr: bytes, relation_id: int, requirer_csrs: list[RequirerCSR
         return True
 
 
+class AllowedFields:
+    """Filter the CSR so as to only allow CSRs that match the given regexes for the CSR fields."""
+
+    EMAIL_SUBJECT_NAME = "1.2.840.113549.1.9.1"
+    ORGANIZATION_SUBJECT_NAME = "O"
+    COUNTRY_CODE_SUBJECT_NAME = "C"
+
+    def __init__(self, filters: dict):
+        self.field_filters = filters
+
+    def evaluate(self, csr: bytes, relation_id: int, requirer_csrs: list[RequirerCSR]) -> bool:  # noqa: C901
+        """Accept CSR only if the given CSR passes the field regex matches."""
+        csr_object = x509.load_pem_x509_csr(csr)
+        san = csr_object.extensions.get_extension_for_class(x509.SubjectAlternativeName).value
+        subject = csr_object.subject
+        if challenge := self.field_filters.get("allowed-dns"):
+            pattern = re.compile(challenge)
+            dns_list = san.get_values_for_type(x509.DNSName)
+            for dns in dns_list:
+                if not pattern.match(dns):
+                    logger.warning("DNS validation failed in csr from relation id %s", relation_id)
+                    return False
```

Comment:
I think the subject could contain a DNS also and it does not seem to be checked here. It can probably also contain an OID or IP.

Reply:
https://www.rfc-editor.org/rfc/rfc5280#section-4.1 

Looks like the fields that contain distinguished names are: issuer, subject, and subjectAltName extensions. So the filter needs to look at subject and san like you said.

Also, email dn in the subject seems to be deprecated but permitted.

There is also a [name constraints extension](https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.10) defined here, which must be used by a CA which will apply to subject and subject alternative names. This is also not relevant.